origin_isolation.md

```diff
@@ -0,0 +1,25 @@
+
+If the main application embeds a hidden frame within itself, this hidden frame can load the alternative origin application. One frame can pass a message to another frame using the postMessage function, even if the frames are not hosted by the same origin. This is the fundamental feature that allows the main application to make access token-bearing requests to resource servers without having the ability to directly read the access token; instead, requests are initiated by sending a message using postMessage to the alternative-origin application frame. The alternative-origin application makes the token-bearing request to the resource server; the response is then passed back to the main application as another message.
+
+For this technique to be practical to implement, there should be a mechanism for transparently intercepting network requests to resource servers made by the main application. These intercepted requests would then be forwarded to the alternative origin frame, and the responses from the frame would then be transparently presented back to the application. By having this interception mechanism, application developers could continue using the normal network request techniques that they normally use; there would be no need to alter the application substantially in order to directly make postMessage requests. This mechanism for transparent request interception and response can be thought of as an "Identity Proxy".
+
+An Identity Proxy can be built using different techniques. For more modern browsers, there is a feature called a "service worker" which is particularly well-suited for this purpose. A service workers is a separate JavaScript thread which exists in the browser apart from the primary application thread. The browser will pass any network request issued by the primary application to the service worker, giving the service worker an opportunity to alter the request or the response that the primary application receives. By programming the service worker to use postMessage to the alternate-origin frame, it can provide the sort of secure, transparent access token usage that is the goal of an Identity Proxy.
+
+For browsers which do not support service workers, there is another way to build an Identity Proxy. The mechanism used by older browsers to make network requests is called the "XMLHttpRequest" object. The behavior of this object can be customized by using JavaScript to override its default methods. Customizing it so that network requests to resource servers result in postMessage calls to the alternate-origin frame achieves the same result as the service worker method.
+
+Since the acquisition and usage of access tokens are well-defined parts of the OAuth 2 standard, this is a design pattern that can be packaged into a reusable library and shared amongst any browser-based OAuth 2 client application. Doing so allows the application developer to quickly and easily meet their obligation to keep access token values secure.
```

Comment:
This file doesn't seem to be referenced from the main README, is that intentional?

Reply:
I also have a thought in regard to the origin isolation topic, but it may deserve a separate discussion/PR.